internal/app/go-filecoin/storage_market_connector/client.go

```diff
@@ -0,0 +1,229 @@
+	head := s.chainStore.Head()
+	var spState spapow.StoragePowerActorState
+	err := s.chainStore.GetActorStateAt(ctx, head, fcaddr.StoragePowerAddress, &spState)
+	if err != nil {
+		return nil, err
+	}
+
+	infos := []*storagemarket.StorageProviderInfo{}
+	powerHamt, err := hamt.LoadNode(ctx, s.cborStore, spState.PowerTable)
+	if err != nil {
+		return nil, err
+	}
+
+	err = powerHamt.ForEach(ctx, func(minerAddrStr string, _ interface{}) error {
+		fcMinerAddr, err := fcaddr.NewFromString(minerAddrStr)
+		if err != nil {
+			return err
+		}
+
+		var mState spaminer.StorageMinerActorState
+		err = s.chainStore.GetActorStateAt(ctx, head, fcMinerAddr, &mState)
+		if err != nil {
+			return err
+		}
+
+		minerAddr, err := address.NewFromString(minerAddrStr)
+		if err != nil {
+			return err
+		}
+
+		info := mState.Info
+		infos = append(infos, &storagemarket.StorageProviderInfo{
+			Address:    minerAddr,
+			Owner:      info.Owner,
+			Worker:     info.Worker,
+			SectorSize: uint64(info.SectorSize),
+			PeerID:     info.PeerId,
+		})
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return infos, nil
+}
+
+// ValidatePublishedDeal validates a deal has been published correctly
+// Adapted from https://github.com/filecoin-project/lotus/blob/3b34eba6124d16162b712e971f0db2ee108e0f67/markets/storageadapter/client.go#L156
+func (s *StorageClientNodeConnector) ValidatePublishedDeal(ctx context.Context, deal storagemarket.ClientDeal) (uint64, error) {
+	// Fetch receipt to return dealId
+	waitCtx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
+	defer cancel()
+
+	var publishMsg *types.SignedMessage
+	var publishReceipt *types.MessageReceipt
+
+	err := s.waiter.Wait(waitCtx, *deal.PublishMessage, func(block *block.Block, signedMessage *types.SignedMessage, receipt *types.MessageReceipt) error {
+		publishMsg = signedMessage
+		publishReceipt = receipt
+		return nil
+	})
+
+	if err != nil {
+		return 0, err
+	}
+
+	unsigned := publishMsg.Message
+
+	minerWorker, err := s.getFCWorker(ctx, deal.Proposal.Provider)
+	if err != nil {
+		return 0, err
+	}
+
+	if unsigned.From != minerWorker {
+		return 0, xerrors.Errorf("deal wasn't published by storage provider: from=%s, provider=%s", unsigned.From, deal.Proposal.Provider)
+	}
+
+	if unsigned.To != fcaddr.StorageMarketAddress {
+		return 0, xerrors.Errorf("deal publish message wasn't set to StorageMarket actor (to=%s)", unsigned.To)
+	}
+
+	if unsigned.Method != fcsm.PublishStorageDeals {
+		return 0, xerrors.Errorf("deal publish message called incorrect method (method=%s)", unsigned.Method)
+	}
+
+	values, err := abi.DecodeValues(unsigned.Params, []abi.Type{abi.StorageDealProposals})
+	if err != nil {
+		return 0, err
+	}
+
+	msgProposals := values[0].Val.([]types.StorageDealProposal)
+
+	proposal := msgProposals[0] // TODO: Support more than one deal
+
+	// TODO: find a better way to do this
```

Comment:
Would `reflect.DeepEqual` help here?  If not seems like writing a custom `Equals` or `CompareForValidation` method on proposal would be nice.  Seems like it shouldn't have to wait for after this PR.